lnrpc/routerrpc/router_server.go

```diff
@@ -373,52 +380,120 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
 
-	if len(req.Dest) != 33 {
-		return nil, errors.New("invalid length destination key")
-	}
-	var destNode route.Vertex
-	copy(destNode[:], req.Dest)
-
-	// Next, we'll convert the amount in satoshis to mSAT, which are the
-	// native unit of LN.
-	amtMsat := lnwire.NewMSatFromSatoshis(btcutil.Amount(req.AmtSat))
-
-	// Pick a fee limit
-	//
-	// TODO: Change this into behaviour that makes more sense.
-	feeLimit := lnwire.NewMSatFromSatoshis(btcutil.SatoshiPerBitcoin)
-
-	// Finally, we'll query for a route to the destination that can carry
-	// that target amount, we'll only request a single route. Set a
-	// restriction for the default CLTV limit, otherwise we can find a route
-	// that exceeds it and is useless to us.
-	mc := s.cfg.RouterBackend.MissionControl
-	routeReq, err := routing.NewRouteRequest(
-		s.cfg.RouterBackend.SelfNode, &destNode, amtMsat, 0,
-		&routing.RestrictParams{
-			FeeLimit:          feeLimit,
-			CltvLimit:         s.cfg.RouterBackend.MaxTotalTimelock,
-			ProbabilitySource: mc.GetProbability,
-		}, nil, nil, nil, s.cfg.RouterBackend.DefaultFinalCltvDelta,
+	if len(req.PaymentRequest) == 0 {
+		return nil, errors.New("must specify an invoice")
+	}
+
+	payReq, err := zpay32.Decode(
+		req.PaymentRequest, s.cfg.RouterBackend.ActiveNetParams,
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	route, _, err := s.cfg.Router.FindRoute(routeReq)
-	if err != nil {
+	// Generate random payment hash, so we can be sure that the target of
+	// the probe payment doesn't have the preimage to settle the htlc.
+	var paymentHash lntypes.Hash
+	if _, err := crand.Read(paymentHash[:]); err != nil {
+		log.Error("Cannot generate random preimage: %v", err)
 		return nil, err
 	}
 
-	return &RouteFeeResponse{
-		RoutingFeeMsat: int64(route.TotalFees()),
-		TimeLockDelay:  int64(route.TotalTimeLock),
-	}, nil
+	request := &SendPaymentRequest{
+		TimeoutSeconds:   req.TimeoutSeconds,
+		MaxParts:         1,
+		AllowSelfPayment: false,
+		PaymentHash:      paymentHash[:],
+		Dest:             payReq.Destination.SerializeCompressed(),
+		AmtMsat:          int64(*payReq.MilliSat),
+		FeeLimitSat:      100_000_000,
+		PaymentAddr:      payReq.PaymentAddr[:],
+		RouteHints: invoicesrpc.CreateRPCRouteHints(
+			payReq.RouteHints,
+		),
+	}
+
+	return s.sendProbePayment(ctx, request)
+}
+
+// sendProbePayment sends a payment to a target node in order to obtain
+// potential routing fees for it. The payment request has to contain a payment
+// hash that is guaranteed to be unknown to the target node, so it cannot settle
+// the payment. This method invokes a payment request loop in a goroutine and
+// awaits payment status updates.
+func (s *Server) sendProbePayment(ctx context.Context,
+	req *SendPaymentRequest) (*RouteFeeResponse, error) {
+
+	// We'll launch a goroutine to send the payment probes.
+	errChan := make(chan error, 1)
+	defer close(errChan)
+	go func() {
+		err := s.SendPaymentV2(req, nil)
```

Comment:
I'm not sure if passing nil in here is the best solution. There is only a single `probeResultChan`, it gets reused for different `EstimateRouteFee` calls, so messages could interfere?

Perhaps passing in a stream implementation and then consume from it could work. I quickly tried it and the itest passed. Not sure if more of the `Router_SendPaymentV2Server` needs to be implemented. That way also the context could be reused.

```go
// probePaymentStream is a custom implementation of the grpc.ServerStream
// interface. It is used to send payment status updates to the caller on the
// stream channel.
type probePaymentStream struct {
	Router_SendPaymentV2Server

	stream chan *lnrpc.Payment
	ctx    context.Context
}

// Implement the grpc.ServerStream interface for probePaymentStream.
func (p *probePaymentStream) Send(response *lnrpc.Payment) error {
	p.stream <- response

	return nil
}

// Context returns the context of the stream.
func (p *probePaymentStream) Context() context.Context {
	return p.ctx
}

```